File: src/lab5C/ffs_imp.c
#include <stdio.h>
#include <stdlib.h>

typedef unsigned int uint32_t;

uint32_t ffs_imp(uint32_t i) {
    char n = 1;
    if (!(i & 0xffff)) { n += 16; i >>= 16; }
    if (!(i & 0x00ff)) { n += 8;  i >>= 8; }
    if (!(i & 0x000f)) { n += 4;  i >>= 4; }
    if (!(i & 0x0003)) { n += 2;  i >>= 2; }
    return (i) ? (n+((i+1) & 0x01)) : 0;
}

int main (int argc, char **argv) {
   if (argc != 2) {
      printf("Usage: %s <number>\n", argv[0]);
      exit(0);
   }
   uint32_t n = atol(argv[1]);
   printf("ffs_imp: first 1 at %d\n", ffs_imp(n));
}

File: src/lab5C/ffs_imp.java
// compile with javac -g FFS.java (-g generates all debug information)
public class ffs_imp {
   static int ffs_imp(int i) {
      byte n = 1;
      if ((i & 0xffff) == 0) { n += 16; i >>= 16; }
      if ((i & 0x00ff) == 0) { n +=  8; i >>=  8; }
      if ((i & 0x000f) == 0) { n +=  4; i >>=  4; }
      if ((i & 0x0003) == 0) { n +=  2; i >>=  2; }
      if (i != 0) { return (n+((i+1) & 0x01)); } else { return 0; }
   }
   
   public static void main (String args[]) {
      int k = Integer.parseInt(args[0]);
      System.out.println("First 1 is at "+ffs_imp(k));		
   }
}
